src/NCPkgFilterService.cc

```diff
@@ -0,0 +1,310 @@
+
+   return tag;
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::getService
+//	METHOD TYPE : ZyppService
+//
+//	DESCRIPTION : Get service reference from selected line's tag
+//
+
+ZyppService NCPkgServiceTable::getService( int index )
+{
+    NCPkgServiceTag *t = getTag( index );
+
+    if ( !t )
+    {
+	return ZyppService( );
+    }
+    else
+    {
+	 return t->getService();
+    }
+}
+
+static std::string html_escape(const std::string& s)
+{
+    std::string escaped = boost::replace_all_copy(s, "&", "&amp;");
+    boost::replace_all(escaped, "<", "&lt;");
+    boost::replace_all(escaped, ">", "&gt;");
+    return escaped;
+}
+
+std::string NCPkgServiceTable::getDescription(ZyppService r)
+{
+    zypp::ServiceInfo si = repo_manager->getService(r);
+
+    std::string label = _( "<b>Service URL:</b>" );
+    std::string ret = label + html_escape(si.url().asString());
+    return ret;
+}
+
+/////////////////////////////////////////////////////////////////////
+////
+////
+////	METHOD NAME : NCPkgFilterService::fillServiceList
+////	METHOD TYPE : bool
+////
+////	DESCRIPTION : Add items to the service list (assoc.
+////		      product name, if any, and URL)
+////
+//
+bool NCPkgServiceTable::fillServiceList()
+{
+    yuiMilestone() << "Filling service list" << endl;
+
+    std::set<std::string> seen_services;
+
+    std::for_each(ZyppRepositoriesBegin(), ZyppRepositoriesEnd(), [&](const zypp::Repository& repo) {
+        const std::string &service_name(repo.info().service());
+        if (!service_name.empty())
+        {
+            if (seen_services.find(service_name) == seen_services.end())
+            {
+              seen_services.insert(service_name);
+
+              std::vector <std::string> oneLine;
+              oneLine.push_back( service_name );
+              addLine( service_name, oneLine);
+            }
+        }
+    });
+
+    return true;
+}
+
+void NCPkgServiceTable::showServicePackages( )
+{
+    int index = getCurrentItem();
+    ZyppService service = getService( index );
+
+    yuiMilestone() << "Selected service " << service << endl;
+    yuiMilestone() << "Collecting packages in selected service" << endl;
+
+    NCPkgTable *pkgList = packager->PackageList();
+    //clean the pkg table first
+    pkgList->itemsCleared ();
+
+    zypp::PoolQuery query;
+    query.addKind( zypp::ResKind::package );
+
+    std::for_each(ZyppRepositoriesBegin(), ZyppRepositoriesEnd(), [&](const zypp::Repository& repo) {
+        if (service == repo.info().service())
+        {
+          yuiMilestone() << "Adding repo filter: " << repo.info().alias() << std::endl;
+          query.addRepo( repo.info().alias() );
+        }
+    });
+
+    for( zypp::PoolQuery::Selectable_iterator it = query.selectableBegin();
+	it != query.selectableEnd(); it++)
+    {
+        ZyppPkg pkg = tryCastToZyppPkg( (*it)->theObj() );
+        pkgList->createListEntry ( pkg, *it);
```

Comment:
Not sure if this can happen in real life, but I guess that `tryCastToZyppPkg` can also fail (thus the _try_ prefix), and then it will probably return 0 or an invalid `ZyppPkg` which probably can't be used for `pkgList->createListEntry()`. Please check.

Reply:
OK then.